CSS Transforms Module Improvements (#30783)
mdn/content: files/en-us/web/css/css_transforms/index.md

title: CSS transforms
slug: Web/CSS/CSS_transforms
page-type: css-module

> spec-urls: https://drafts.csswg.org/css-transforms-2/

Reviewer: ```suggestion
spec-urls: 
  - https://drafts.csswg.org/css-transforms/
  - https://drafts.csswg.org/css-transforms-2/
```

You can also use the `perspective` slider to modify the [`perspective`](/en-US/docs/Web/CSS/perspective) property of the cube's container, which determines the distance between you and the `z=0` plane.

The [`perspective-origin`](/en-US/docs/Web/CSS/perspective-origin) sliders determine where you, the viewer, are looking into the 3D space for purposes of determining the view's _vanishing point_. This vanishing point is indicated by a small red dot. You can imagine modifying these sliders as physically moving your head up, down, left, and right to see different parts of the cube without moving the cube itself.

The `backface-visibility` checkbox determines whether the cube's back faces are set to `visible` or `hidden`.

The cube in the above example is comprised of six `<div>` elements, all of which are styled with CSS to create the cube's faces. The cube is not drawn using a 2D or 3D canvas context, so **you can inspect the faces of the cube with your browser's developer tools as you would inspect any other DOM element**. Try using your browser's developer tools element picker to inspect different faces of the cube as you transform its position and rotation.

> > **Note:** The order in which 3D rotations are applied affects the resultant transformation. In the above example, rotations are applied in the order X → Y → Z.

Reviewer: ```suggestion
> **Note:** The order in which transformations, including 3D rotations, are applied affects the resultant transformation. In the above example, transforms are translated, scaled, rotated, then skewed. The rotations are applied in the order X → Y → Z.
```

> > **Note:** that the order in which 3D rotations are applied affects the resultant transformation. In the above example, rotations are applied in the order X → Y → Z.

Reviewer: ```suggestion
> **Note:** The order in which transformations, including 3D rotations, are applied affects the resultant transformation. In the above example, transforms are translated, scaled, rotated, then skewed. The rotations are applied in the order X → Y → Z.
```

Author: thank you for your suggestion. in the future, i would appreciate an opportunity to review suggestions before you commit them and/or merge the PR (except in cases where we'd talked about that being OK previously).

Reviewer: Will do.

- {{cssxref("perspective-origin")}}
- {{cssxref("rotate")}}
- {{cssxref("scale")}}

> - [`skew`](/en-US/docs/Web/CSS/transform-function/skew)

Reviewer: this is not a property. it's a function.

The only properties defined in this module are transform, transform-box, and transform-origin. The spec needs to be updated to https://drafts.csswg.org/css-transforms-2 instead. 

this list should be:

perspective
perspective-origin
rotate
scale
transform
t-box
t-origin
t-style
translate
transform
transform-origin
transformable element
transformation matrix
transformed element

- {{cssxref("transform")}}
- {{cssxref("transform-box")}}
- {{cssxref("transform-origin")}}
- {{cssxref("transform-style")}}
- {{cssxref("translate")}}

### Functions

> - See {{cssxref("&lt;transform-function&gt;")}}

Reviewer: Every function needs to all be listed here. CSS module landing page links to all the features of the specification. Everything in the spec should presented as a bulleted list under the relevant headings.

```suggestion
- {{cssxref("matrix()")}}
- {{cssxref("rotate()")}}
- {{cssxref("scale()")}}
- {{cssxref("scaleX()")}}
- {{cssxref("scaleY()")}}
- {{cssxref("skew()")}}
- {{cssxref("skewX()")}}
- {{cssxref("skewY()")}}
- {{cssxref("translate()")}}
- {{cssxref("translateX()")}}
- {{cssxref("translateY()")}}
```
then add the functions listed here:

matrix3d() = matrix3d( [<number>](https://drafts.csswg.org/css-values-4/#number-value)[#{16}](https://drafts.csswg.org/css-values-4/#mult-comma) )
specifies a 3D transformation as a 4x4 homogeneous matrix of 16 values in column-major order.

translate3d() = translate3d( [<length-percentage>](https://drafts.csswg.org/css-values-4/#typedef-length-percentage) [,](https://drafts.csswg.org/css-values-4/#comb-comma) <length-percentage> , [<length>](https://drafts.csswg.org/css-values-4/#length-value) )
specifies a [3D translation](https://drafts.csswg.org/css-transforms-2/#Translate3dDefined) by the vector [tx,ty,tz], with tx, ty and tz being the first, second and third translation-value parameters respectively.

translateZ() = translateZ( [<length>](https://drafts.csswg.org/css-values-4/#length-value) )
specifies a [3D translation](https://drafts.csswg.org/css-transforms-2/#Translate3dDefined) by the vector [0,0,tz] with the given amount in the Z direction.

scale3d() = scale3d( [ [<number>](https://drafts.csswg.org/css-values-4/#number-value) [|](https://drafts.csswg.org/css-values-4/#comb-one) [<percentage>](https://drafts.csswg.org/css-values-4/#percentage-value) ][#{3}](https://drafts.csswg.org/css-values-4/#mult-comma) )
specifies a [3D scale](https://drafts.csswg.org/css-transforms-2/#Scale3dDefined) operation by the [sx,sy,sz] scaling vector described by the 3 parameters.

scaleZ() = scaleZ( [ [<number>](https://drafts.csswg.org/css-values-4/#number-value) [|](https://drafts.csswg.org/css-values-4/#comb-one) [<percentage>](https://drafts.csswg.org/css-values-4/#percentage-value) ] )
specifies a [3D scale](https://drafts.csswg.org/css-transforms-2/#Scale3dDefined) operation using the [1,1,sz] scaling vector, where sz is given as the parameter.

rotate3d() = rotate3d( [<number>](https://drafts.csswg.org/css-values-4/#number-value) [,](https://drafts.csswg.org/css-values-4/#comb-comma) <number> , <number> , [ [<angle>](https://drafts.csswg.org/css-values-4/#angle-value) [|](https://drafts.csswg.org/css-values-4/#comb-one) [<zero>](https://drafts.csswg.org/css-values-4/#zero-value) ] )
specifies a [3D rotation](https://drafts.csswg.org/css-transforms-2/#Rotate3dDefined) by the angle specified in last parameter about the [x,y,z] direction vector described by the first three parameters. A direction vector that cannot be normalized, such as [0,0,0], will cause the rotation to not be applied.

NOTE: the rotation is clockwise as one looks from the end of the vector toward the origin.

rotateX() = rotateX( [ [<angle>](https://drafts.csswg.org/css-values-4/#angle-value) [|](https://drafts.csswg.org/css-values-4/#comb-one) [<zero>](https://drafts.csswg.org/css-values-4/#zero-value) ] )
same as rotate3d(1, 0, 0, <angle>).

rotateY() = rotateY( [ [<angle>](https://drafts.csswg.org/css-values-4/#angle-value) [|](https://drafts.csswg.org/css-values-4/#comb-one) [<zero>](https://drafts.csswg.org/css-values-4/#zero-value) ] )
same as rotate3d(0, 1, 0, <angle>).

rotateZ() = rotateZ( [ [<angle>](https://drafts.csswg.org/css-values-4/#angle-value) [|](https://drafts.csswg.org/css-values-4/#comb-one) [<zero>](https://drafts.csswg.org/css-values-4/#zero-value) ] )
same as rotate3d(0, 0, 1, <angle>), which is a 3d transform equivalent to the 2d transform rotate(<angle>).

perspective()

we also need the SVG `transform` attribute.

- {{cssxref("rotate")}}
- {{cssxref("scale")}}
- {{cssxref("transform")}}

> - {{cssxref("transform-box")}} _(Defined in [Level 1](https://drafts.csswg.org/css-transforms-1/))_
> - {{cssxref("transform-origin")}} _(Defined in [Level 1](https://drafts.csswg.org/css-transforms-1/))_

Reviewer: ```suggestion
- {{cssxref("transform-box")}}
- {{cssxref("transform-origin")}}
```

- [Using CSS transforms](/en-US/docs/Web/CSS/CSS_transforms/Using_CSS_transforms)
  - : Step-by-step tutorial about how to transform elements styled with CSS.
- [Styling animated buttons with CSS transforms](/en-US/docs/Web/CSS/CSS_transforms/styling_animated_buttons_with_css_transforms)
  - : Contains several examples of `<button>` elements styled using animated CSS transforms.

## Related concepts

> 

Reviewer: Please take a look at the module landing pages that have been updated and conform to how those are presented. 

These are useful links, but should be presented within guides, related concepts as list items under their module as a header, or in See also.

There are a ton of related concepts not listed. A good place to figure out what to include is the

- [Using CSS transforms](/en-US/docs/Web/CSS/CSS_transforms/Using_CSS_transforms)
  - : Step-by-step tutorial about how to transform elements styled with CSS.
- [Coordinate systems](/en-US/docs/Web/CSS/CSSOM_view/Coordinate_systems)
  - : Describes the way pixel locations are defined in the CSS object model.

> - [Matrix math for the web](/en-US/docs/Web/API/WebGL_API/Matrix_math_for_the_web)

Reviewer: ```suggestion
- [Perfomance fundamentals: Using CSS transforms](/en-US/docs/Web/Performance/Fundamentals#use_css_transforms)
   - : An overview of web performance fundamentals, including how CSS tranforms can help.
- [Matrix math for the web](/en-US/docs/Web/API/WebGL_API/Matrix_math_for_the_web)
```

  - : Describes how object transformations can be represented by mathematical matrices.

## Related concepts

- CSS Properties:

>   - [`background-position`](/en-US/docs/Web/CSS/background-position)

Reviewer: ```suggestion
  - [`animation`](/en-US/docs/Web/CSS/animation)
  - [`background-position`](/en-US/docs/Web/CSS/background-position)
```

  - [`clip`](/en-US/docs/Web/CSS/clip)
  - [`clip-path`](/en-US/docs/Web/CSS/clip-path)
  - [`contain`](/en-US/docs/Web/CSS/contain)
  - [`content-visibility`](/en-US/docs/Web/CSS/content-visibility)
  - [`isolation`](/en-US/docs/Web/CSS/isolation)
  - [`mask`](/en-US/docs/Web/CSS/mask)
  - [`mask-border-source`](/en-US/docs/Web/CSS/mask-border-source)
  - [`mask-image`](/en-US/docs/Web/CSS/mask-image)
  - [`mix-blend-mode`](/en-US/docs/Web/CSS/mix-blend-mode)
  - [`opacity`](/en-US/docs/Web/CSS/opacity)
  - [`overflow`](/en-US/docs/Web/CSS/overflow)

>   - [`visibility`](/en-US/docs/Web/CSS/visibility)

Reviewer: ```suggestion
  - [`transition`](/en-US/docs/Web/CSS/transition)
  - [`visibility`](/en-US/docs/Web/CSS/visibility)
```

- Data types:
  - [`<angle>`](/en-US/docs/Web/CSS/angle)
  - [`<length-percentage>`](/en-US/docs/Web/CSS/length-percentage)
  - [`<length>`](/en-US/docs/Web/CSS/length)
  - [`<number>`](/en-US/docs/Web/CSS/number)
  - [`<percentage>`](/en-US/docs/Web/CSS/percentage)
  - [`<position>`](/en-US/docs/Web/CSS/position_value)
- Glossary terms:
  - [Interpolation](/en-US/docs/Glossary/Interpolation)
  - [Stacking context](/en-US/docs/Glossary/Stacking_context)
- SVG concepts:

>   - [`<animate>`](/en-US/docs/Web/SVG/Element/animate)
>   - [`<set>`](/en-US/docs/Web/SVG/Element/set)
>   - [`transform`](/en-US/docs/Web/SVG/Attribute/transform)

Reviewer: ```suggestion
  - [`<animate>`](/en-US/docs/Web/SVG/Element/animate) element
  - [`<set>`](/en-US/docs/Web/SVG/Element/set) element
  - [`transform`](/en-US/docs/Web/SVG/Attribute/transform) attribute
```

Reviewer: did you want to add https://developer.mozilla.org/en-US/docs/Web/SVG/Element/animateTransform?

Author: sure!

## Specifications

{{Specifications}}

## See also

> - CSS properties: [`animation`](/en-US/docs/Web/CSS/animation), [`transition`](/en-US/docs/Web/CSS/transition)

Reviewer: ```suggestion
- [CSS animations](/en-US/docs/Web/CSS/CSS_animations) module
- [CSS transitions](/en-US/docs/Web/CSS/CSS_transitions) module
```

Reviewer: let's also link to SVG transforms overview, if there is one, and maybe link the SVG in "SVG concepts" to the svg landing page.
